#!/usr/bin/env python3
"""
ÐŸÐ¾Ð»Ð½Ñ‹Ð¹ Ñ‚ÐµÑÑ‚ gRPC Ð·Ð°Ð¿Ñ€Ð¾ÑÐ° Ñ Ð¿Ð¾Ð»ÑƒÑ‡ÐµÐ½Ð¸ÐµÐ¼ Ð¾Ñ‚Ð²ÐµÑ‚Ð°
"""
import asyncio
import grpc
from grpc import aio
import sys
import os
import subprocess
import uuid

# Ð”Ð¾Ð±Ð°Ð²Ð»ÑÐµÐ¼ Ð¿ÑƒÑ‚ÑŒ Ðº protobuf Ñ„Ð°Ð¹Ð»Ð°Ð¼
sys.path.append(os.path.join(os.path.dirname(__file__), '../modules/grpc_service'))
try:
    import streaming_pb2
    import streaming_pb2_grpc
except ImportError as e:
    print(f"âŒ ÐžÑˆÐ¸Ð±ÐºÐ° Ð¸Ð¼Ð¿Ð¾Ñ€Ñ‚Ð° protobuf: {e}")
    sys.exit(1)

def get_server_certificate(host: str, port: int) -> bytes:
    """ÐŸÐ¾Ð»ÑƒÑ‡ÐµÐ½Ð¸Ðµ ÑÐµÑ€Ñ‚Ð¸Ñ„Ð¸ÐºÐ°Ñ‚Ð° ÑÐµÑ€Ð²ÐµÑ€Ð°"""
    try:
        result = subprocess.run(
            ['openssl', 's_client', '-connect', f'{host}:{port}', '-showcerts'],
            input=b'', capture_output=True, timeout=5
        )
        if result.returncode == 0:
            cert_start = result.stdout.find(b'-----BEGIN CERTIFICATE-----')
            cert_end = result.stdout.find(b'-----END CERTIFICATE-----', cert_start)
            if cert_start != -1 and cert_end != -1:
                return result.stdout[cert_start:cert_end + len(b'-----END CERTIFICATE-----')]
    except Exception as e:
        print(f"âš ï¸  ÐžÑˆÐ¸Ð±ÐºÐ° Ð¿Ð¾Ð»ÑƒÑ‡ÐµÐ½Ð¸Ñ ÑÐµÑ€Ñ‚Ð¸Ñ„Ð¸ÐºÐ°Ñ‚Ð°: {e}")
    return None

async def test_full_request(host: str = "nexy-server.canadacentral.cloudapp.azure.com", port: int = 443):
    """ÐŸÐ¾Ð»Ð½Ñ‹Ð¹ Ñ‚ÐµÑÑ‚ gRPC Ð·Ð°Ð¿Ñ€Ð¾ÑÐ°"""
    address = f"{host}:{port}"
    print("=" * 70)
    print("ðŸ” ÐŸÐžÐ›ÐÐ«Ð™ Ð¢Ð•Ð¡Ð¢ gRPC Ð—ÐÐŸÐ ÐžÐ¡Ð")
    print("=" * 70)
    print()
    
    try:
        # Ð¡Ð¾Ð·Ð´Ð°Ñ‘Ð¼ secure_channel Ð´Ð»Ñ Ð¿Ð¾Ñ€Ñ‚Ð° 443
        print(f"ðŸ”’ Ð¡Ð¾Ð·Ð´Ð°Ð½Ð¸Ðµ secure_channel Ðº {address}...")
        cert_pem = get_server_certificate(host, port)
        
        if cert_pem:
            credentials = grpc.ssl_channel_credentials(root_certificates=cert_pem)
            print("âœ… Ð¡ÐµÑ€Ñ‚Ð¸Ñ„Ð¸ÐºÐ°Ñ‚ ÑÐµÑ€Ð²ÐµÑ€Ð° Ð¿Ð¾Ð»ÑƒÑ‡ÐµÐ½")
        else:
            credentials = grpc.ssl_channel_credentials()
            print("âš ï¸  Ð˜ÑÐ¿Ð¾Ð»ÑŒÐ·ÑƒÑŽÑ‚ÑÑ ÑÑ‚Ð°Ð½Ð´Ð°Ñ€Ñ‚Ð½Ñ‹Ðµ credentials")
        
        channel = aio.secure_channel(address, credentials)
        stub = streaming_pb2_grpc.StreamingServiceStub(channel)
        
        # ÐŸÑ€Ð¾Ð²ÐµÑ€ÑÐµÐ¼ Ð³Ð¾Ñ‚Ð¾Ð²Ð½Ð¾ÑÑ‚ÑŒ ÐºÐ°Ð½Ð°Ð»Ð°
        try:
            await asyncio.wait_for(channel.channel_ready(), timeout=10.0)
            print("âœ… ÐšÐ°Ð½Ð°Ð» Ð³Ð¾Ñ‚Ð¾Ð² Ðº Ð¸ÑÐ¿Ð¾Ð»ÑŒÐ·Ð¾Ð²Ð°Ð½Ð¸ÑŽ")
        except asyncio.TimeoutError:
            print("âš ï¸  Ð¢Ð°Ð¹Ð¼Ð°ÑƒÑ‚ Ð³Ð¾Ñ‚Ð¾Ð²Ð½Ð¾ÑÑ‚Ð¸ ÐºÐ°Ð½Ð°Ð»Ð° (Ð¿Ñ€Ð¾Ð´Ð¾Ð»Ð¶Ð°ÐµÐ¼...)")
        
        # Ð¡Ð¾Ð·Ð´Ð°Ñ‘Ð¼ Ð·Ð°Ð¿Ñ€Ð¾Ñ
        print()
        print("ðŸ“¤ ÐžÑ‚Ð¿Ñ€Ð°Ð²ÐºÐ° Ð·Ð°Ð¿Ñ€Ð¾ÑÐ°...")
        session_id = str(uuid.uuid4())
        request = streaming_pb2.StreamRequest(
            prompt="ÐŸÑ€Ð¸Ð²ÐµÑ‚! Ð­Ñ‚Ð¾ Ñ‚ÐµÑÑ‚Ð¾Ð²Ñ‹Ð¹ Ð·Ð°Ð¿Ñ€Ð¾Ñ. Ð Ð°ÑÑÐºÐ°Ð¶Ð¸, ÐºÐ°Ðº Ð´ÐµÐ»Ð°?",
            hardware_id="test-device-full-12345",
            session_id=session_id
        )
        print(f"   Session ID: {session_id}")
        print(f"   Hardware ID: {request.hardware_id}")
        print(f"   Prompt: {request.prompt}")
        print()
        
        # ÐŸÐ¾Ð»ÑƒÑ‡Ð°ÐµÐ¼ Ð¾Ñ‚Ð²ÐµÑ‚Ñ‹
        print("ðŸ“¥ ÐžÐ¶Ð¸Ð´Ð°Ð½Ð¸Ðµ Ð¾Ñ‚Ð²ÐµÑ‚Ð¾Ð² Ð¾Ñ‚ ÑÐµÑ€Ð²ÐµÑ€Ð°...")
        print("-" * 70)
        
        text_chunks = []
        audio_chunks = []
        end_message = None
        error_message = None
        
        try:
            async for response in stub.StreamAudio(request, timeout=60.0):
                if response.HasField('text_chunk'):
                    text = response.text_chunk
                    text_chunks.append(text)
                    print(f"ðŸ“ Text chunk: {text[:100]}...")
                
                elif response.HasField('audio_chunk'):
                    audio_data = response.audio_chunk.audio_data
                    audio_chunks.append(audio_data)
                    print(f"ðŸ”Š Audio chunk: {len(audio_data)} Ð±Ð°Ð¹Ñ‚")
                
                elif response.HasField('end_message'):
                    end_message = response.end_message
                    print(f"âœ… End message: {end_message}")
                    break
                
                elif response.HasField('error_message'):
                    error_message = response.error_message
                    print(f"âŒ Error message: {error_message}")
                    break
            
            print("-" * 70)
            print()
            print("ðŸ“Š Ð Ð•Ð—Ð£Ð›Ð¬Ð¢ÐÐ¢Ð«:")
            print(f"   Text chunks: {len(text_chunks)}")
            if text_chunks:
                full_text = "".join(text_chunks)
                print(f"   ÐŸÐ¾Ð»Ð½Ñ‹Ð¹ Ñ‚ÐµÐºÑÑ‚ ({len(full_text)} ÑÐ¸Ð¼Ð²Ð¾Ð»Ð¾Ð²): {full_text[:200]}...")
            print(f"   Audio chunks: {len(audio_chunks)}")
            if audio_chunks:
                total_audio = sum(len(chunk) for chunk in audio_chunks)
                print(f"   ÐžÐ±Ñ‰Ð¸Ð¹ Ñ€Ð°Ð·Ð¼ÐµÑ€ Ð°ÑƒÐ´Ð¸Ð¾: {total_audio} Ð±Ð°Ð¹Ñ‚")
            if end_message:
                print(f"   âœ… Ð—Ð°Ð²ÐµÑ€ÑˆÐµÐ½Ð¾: {end_message}")
            if error_message:
                print(f"   âŒ ÐžÑˆÐ¸Ð±ÐºÐ°: {error_message}")
            
            # Ð˜Ñ‚Ð¾Ð³Ð¾Ð²Ð°Ñ Ð¾Ñ†ÐµÐ½ÐºÐ°
            print()
            if error_message:
                print("âŒ Ð¢Ð•Ð¡Ð¢ ÐÐ• ÐŸÐ ÐžÐ™Ð”Ð•Ð: Ð¿Ð¾Ð»ÑƒÑ‡ÐµÐ½Ð° Ð¾ÑˆÐ¸Ð±ÐºÐ° Ð¾Ñ‚ ÑÐµÑ€Ð²ÐµÑ€Ð°")
                return False
            elif len(text_chunks) > 0 or len(audio_chunks) > 0:
                print("âœ… Ð¢Ð•Ð¡Ð¢ ÐŸÐ ÐžÐ™Ð”Ð•Ð: Ð¿Ð¾Ð»ÑƒÑ‡ÐµÐ½Ñ‹ Ð´Ð°Ð½Ð½Ñ‹Ðµ Ð¾Ñ‚ ÑÐµÑ€Ð²ÐµÑ€Ð°")
                return True
            else:
                print("âš ï¸  Ð¢Ð•Ð¡Ð¢ Ð§ÐÐ¡Ð¢Ð˜Ð§ÐÐž ÐŸÐ ÐžÐ™Ð”Ð•Ð: Ð¾Ñ‚Ð²ÐµÑ‚ Ð¿Ð¾Ð»ÑƒÑ‡ÐµÐ½, Ð½Ð¾ Ð´Ð°Ð½Ð½Ñ‹Ñ… Ð½ÐµÑ‚")
                return True
        
        except grpc.RpcError as e:
            print(f"âŒ gRPC Ð¾ÑˆÐ¸Ð±ÐºÐ°: {e.code()} - {e.details()}")
            return False
        
        finally:
            await channel.close()
            print("âœ… ÐšÐ°Ð½Ð°Ð» Ð·Ð°ÐºÑ€Ñ‹Ñ‚")
    
    except Exception as e:
        print(f"âŒ ÐšÑ€Ð¸Ñ‚Ð¸Ñ‡ÐµÑÐºÐ°Ñ Ð¾ÑˆÐ¸Ð±ÐºÐ°: {e}")
        import traceback
        traceback.print_exc()
        return False

if __name__ == "__main__":
    host = sys.argv[1] if len(sys.argv) > 1 else "nexy-server.canadacentral.cloudapp.azure.com"
    port = int(sys.argv[2]) if len(sys.argv) > 2 else 443
    
    success = asyncio.run(test_full_request(host, port))
    sys.exit(0 if success else 1)
